uaclient/api/u/pro/services/list/v1.py

```diff
@@ -0,0 +1,217 @@
+        ),
+        Field("status", StringDataValue, doc="Serice Status"),
+    ]
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        desc: str,
+        available: str,
+        entitled: Optional[str] = None,
+        status: Optional[str] = None,
+    ):
+        self.name = name
+        self.description = desc
+        self.available = available
+        self.entitled = entitled
+        self.status = status
+
+
+class ServiceListResult(DataObject, AdditionalInfo):
+    fields = [
+        Field("services", data_list(ServiceInfo), doc="List of services"),
+    ]
+
+    def __init__(self, *, services: List[ServiceInfo]):
+        self.services = services
+
+
+def get_service_list() -> ServiceListResult:
+    return _get_service_list(UAConfig())
+
+
+def _get_service_list(cfg: UAConfig) -> ServiceListResult:
+    if not _is_attached(cfg).is_attached:
+        return unattached_services(cfg)
+    return attached_services(cfg)
+
+
+def attached_services(cfg):
+    services_list = []
+    machine_token_file = machine_token.get_machine_token_file(cfg)
+    resources = machine_token_file.machine_token.get("availableResources")
+    if not resources:
+        resources = get_available_resources(cfg)
+    inapplicable_resources = {
+        resource["name"]: resource.get("description")
+        for resource in sorted(resources, key=lambda x: x.get("name", ""))
+        if not resource.get("available")
+    }
+    for resource in resources:
+        try:
+            ent = entitlement_factory(cfg=cfg, name=resource.get("name", ""))
+        except exceptions.EntitlementNotFoundError:
+            continue
+
+        services_list.append(
+            _get_service_status(cfg, ent, inapplicable_resources)
+        )
+        services_list.sort(key=lambda x: x.name)
+    return ServiceListResult(services=services_list)
+
+
+def unattached_services(cfg):
+    services_list = []
+    resources = get_available_resources(cfg)
+    for resource in resources:
+        if resource.get("available"):
+            available = UserFacingAvailability.AVAILABLE.value
+        else:
+            available = UserFacingAvailability.UNAVAILABLE.value
+        try:
+            ent = entitlement_factory(cfg=cfg, name=resource.get("name", ""))
+        except exceptions.EntitlementNotFoundError:
+            LOG.debug(
+                "Ignoring availability of unknown service %s from contract "
+                "server",
+                resource.get("name", "without a 'name' key"),
+            )
+            continue
+        service = ServiceInfo(
+            name=resource.get("presentedAs", resource["name"]),
+            desc=ent.description,
+            available=available,
+            entitled=None,
+            status=None,
+        )
+        services_list.append(service)
+    return ServiceListResult(services=services_list)
+
+
+def _get_service_status(cfg, ent, inapplicable_resources):
```

Comment:
I know this problem was not introduced in this PR, but I think an entitlement should be able to tell its status. We have logic on different places that check if a service is enabled. I think we should consolidate then on the entitlement class. What do you think @orndorffgrant @renanrodrigo ?

Reply:
agreed, but in a separate changeset imho